### ISSUE TITLE
cupy.ReductionKernel broke from cupy v12 to cupy v13

### ISSUE THREAD

**Issue by McHaillet** [cupy/cupy#8184]
### Description

In our code we use a ReductionKernel to calculate the square sum over a 3D array. This seems to have broken with the update from cupy v12.3.0 to v13.0.0.

### To Reproduce

We use a kernel very similar to what is available in the [cupy docs](https://docs.cupy.dev/en/stable/user_guide/kernel.html#reduction-kernels):
```py
square_sum_kernel = cp.ReductionKernel(
    'T x',  # input params
    'T y',  # output params
    'x * x',  # pre-processing expression
    'a + b',  # reduction operation
    'y = a',  # post-reduction output processing
    '0',  # identity value
    'square sum'  # kernel name
)
x = cp.ones((300,)*3)
xx = square_sum_kernel(x)
```

This fails with the following error:

```
---------------------------------------------------
--- JIT compile log for cupy_jitify_exercise ---
---------------------------------------------------
cub/util_cpp_dialect.cuh(143): warning #161-D: unrecognized #pragma

std/barrier(16): catastrophic error: #error directive: "CUDA synchronization primitives are only supported for sm_70 and up."

1 catastrophic error detected in the compilation of "cupy_jitify_exercise".
Compilation terminated.

---------------------------------------------------
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[9], line 1
----> 1 xx = square_sum_kernel(x)

File cupy/_core/_reduction.pyx:828, in cupy._core._reduction.ReductionKernel.__call__()

File cupy/_core/_reduction.pyx:370, in cupy._core._reduction._AbstractReductionKernel._call()

File cupy/_core/_cub_reduction.pyx:689, in cupy._core._cub_reduction._try_to_call_cub_reduction()

File cupy/_core/_cub_reduction.pyx:526, in cupy._core._cub_reduction._launch_cub()

File cupy/_core/_cub_reduction.pyx:461, in cupy._core._cub_reduction._cub_two_pass_launch()

File cupy/_util.pyx:64, in cupy._util.memoize.decorator.ret()

File cupy/_core/_cub_reduction.pyx:240, in cupy._core._cub_reduction._SimpleCubReductionKernel_get_cached_function()

File cupy/_core/_cub_reduction.pyx:223, in cupy._core._cub_reduction._create_cub_reduction_function()

File cupy/_core/core.pyx:2254, in cupy._core.core.compile_with_cache()

File /data2/mchaillet/programs/tm_gpu_test/env/lib/python3.12/site-packages/cupy/cuda/compiler.py:484, in _compile_module_with_cache(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, jitify)
    480     return _compile_with_cache_hip(
    481         source, options, arch, cache_dir, extra_source, backend,
    482         name_expressions, log_stream, cache_in_memory)
    483 else:
--> 484     return _compile_with_cache_cuda(
    485         source, options, arch, cache_dir, extra_source, backend,
    486         enable_cooperative_groups, name_expressions, log_stream,
    487         cache_in_memory, jitify)

File /data2/mchaillet/programs/tm_gpu_test/env/lib/python3.12/site-packages/cupy/cuda/compiler.py:562, in _compile_with_cache_cuda(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, cache_in_memory, jitify)
    560 if backend == 'nvrtc':
    561     cu_name = '' if cache_in_memory else name + '.cu'
--> 562     ptx, mapping = compile_using_nvrtc(
    563         source, options, arch, cu_name, name_expressions,
    564         log_stream, cache_in_memory, jitify)
    565     if _is_cudadevrt_needed(options):
    566         # for separate compilation
    567         ls = function.LinkState()

File /data2/mchaillet/programs/tm_gpu_test/env/lib/python3.12/site-packages/cupy/cuda/compiler.py:319, in compile_using_nvrtc(source, options, arch, filename, name_expressions, log_stream, cache_in_memory, jitify)
    316         with open(cu_path, 'w') as cu_file:
    317             cu_file.write(source)
--> 319         return _compile(source, options, cu_path,
    320                         name_expressions, log_stream, jitify)
    321 else:
    322     cu_path = '' if not jitify else filename

File /data2/mchaillet/programs/tm_gpu_test/env/lib/python3.12/site-packages/cupy/cuda/compiler.py:284, in compile_using_nvrtc.<locals>._compile(source, options, cu_path, name_expressions, log_stream, jitify)
    280 def _compile(
    281         source, options, cu_path, name_expressions, log_stream, jitify):
    283     if jitify:
--> 284         options, headers, include_names = _jitify_prep(
    285             source, options, cu_path)
    286     else:
    287         headers = include_names = ()

File /data2/mchaillet/programs/tm_gpu_test/env/lib/python3.12/site-packages/cupy/cuda/compiler.py:233, in _jitify_prep(source, options, cu_path)
    231 if not _jitify_header_source_map_populated:
    232     from cupy._core import core
--> 233     jitify._init_module()
    234     jitify._add_sources(core._get_header_source_map())
    235     _jitify_header_source_map_populated = True

File cupy/cuda/jitify.pyx:212, in cupy.cuda.jitify._init_module()

File cupy/cuda/jitify.pyx:233, in cupy.cuda.jitify._init_module()

File cupy/cuda/jitify.pyx:209, in cupy.cuda.jitify._init_cupy_headers()

File cupy/cuda/jitify.pyx:192, in cupy.cuda.jitify._init_cupy_headers_from_scratch()

File cupy/cuda/jitify.pyx:264, in cupy.cuda.jitify.jitify()

RuntimeError: Runtime compilation failed
```

Moreover, these run fine:
```
xx = square_sum_kernel(x, axis=0)
xx = square_sum_kernel(x, axis=1)
xx = square_sum_kernel(x, axis=(0,1))
xx = square_sum_kernel(x, axis=(0,2))
```
But these fail:
```
xx = square_sum_kernel(x, axis=2)
xx = square_sum_kernel(x, axis=(1,2))
```

### Installation

Source (`pip install cupy`)

### Environment

```
OS                           : Linux-4.18.0-372.26.1.el8_6.x86_64-x86_64-with-glibc2.28
Python Version               : 3.12.1
CuPy Version                 : 13.0.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.26.4
SciPy Version                : 1.12.0
Cython Build Version         : 0.29.37
Cython Runtime Version       : None
CUDA Root                    : /opt/apps/cuda-11.8
nvcc PATH                    : /opt/apps/cuda-11.8/bin/nvcc
CUDA Build Version           : 11080
CUDA Driver Version          : 11080
CUDA Runtime Version         : 11080 (linked to CuPy) / 11080 (locally installed)
cuBLAS Version               : (available)
cuFFT Version                : 10900
cuRAND Version               : 10300
cuSOLVER Version             : (11, 4, 1)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 8)
Thrust Version               : 200200
CUB Build Version            : 200200
Jitify Build Version         : <unknown>
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : NVIDIA GeForce GTX 1080 Ti
Device 0 Compute Capability  : 61
Device 0 PCI Bus ID          : 0000:02:00.0
Device 1 Name                : NVIDIA GeForce GTX 1080 Ti
Device 1 Compute Capability  : 61
Device 1 PCI Bus ID          : 0000:03:00.0
Device 2 Name                : NVIDIA GeForce GTX 1080 Ti
Device 2 Compute Capability  : 61
Device 2 PCI Bus ID          : 0000:82:00.0
Device 3 Name                : NVIDIA GeForce GTX 1080 Ti
Device 3 Compute Capability  : 61
Device 3 PCI Bus ID          : 0000:83:00.0
```

### Additional Information

I here show the report for v13, but when forcing installation of v12 with `pip install cupy=12.3.0` the code that I show all runs fine.

**Comment by McHaillet** [comment]
Hi leofang,

Thanks a lot for the quick response! I made a clean install with the suggested line removed and your suspicion was correct, now the ReductionKernel works. 

Just for reference, I made a new python 3 environment, cleaned the cache and ran:
```
git clone --recursive https://github.com/cupy/cupy.git
cd cupy
# removed <cuda/barrier> on line 142 in cupy/cuda/jitify.pyx
python -m pip install .
```
Then ran the code I mentioned above.

Best wishes,
Marten

**Comment by leofang** [comment]
Thanks from bringing this to our attention, @McHaillet. We'll fix it in v13.1.0.

**Comment by McHaillet** [comment]
Thanks a lot for the update!
